_posts/en/newsletters/2019-05-14-newsletter.md

```diff
@@ -0,0 +1,485 @@
+a tree to contain a maximum of around four billion scripts.  However,
+only one of those scripts would appear in any spending transaction and
+only 32 hashes directly related to the merkle tree would need to be
+generated to prove the script was part of the tree---this means that more
+complex scripts than our simple HTLC can gain much larger efficiency
+improvements than seen above (see an article focused on [MAST][] for
+more information).
+
+### Slight changes to scripting with Tapscript
+
+When using Bitcoin Script with Taproot, some of the rules have been
+changed, most notably:
+
+- **Unused and invalid opcodes redefined:** the bytes of opcodes that
+  were never assigned, or which were made invalid by Satoshi Nakamoto,
+  or which were created for upgrades and have not been used yet have all
+  been changed to have the semantics of an `OP_SUCCESS` opcode that
+  unconditionally renders any script containing that code valid.
+  Because future soft forks can only add new rules by making
+  previously-valid things invalid, this maximal validity
+  today will allow maximal flexibility in future soft forks.  Of course,
+  receivers get to choose what scripts they receive payments to, so no
+  one who likes their money will include an `OP_SUCCESSn` opcode in
+  their scripts until a soft fork has given it a new consensus-enforced
+  meaning.
+
+- **Schnorr signatures required:** ECDSA signatures will not be accepted
+  in Tapscript.
+
+- **New script-based multisig semantics:** the current
+  `OP_CHECKMULTISIG` and `OP_CHECKMULTISIGVERIFY` opcodes are not
+  available in Tapscript.  Instead, the existing single-sig
+  `OP_CHECKSIG` and `OP_CHECKSIGVERIFY` may be chained (see example
+  below); or a new `OP_CHECKSIGADD` opcode may be used to increment a
+  counter if a signature matches a specified public key (e.g. for
+  2-of-3, you need two `OP_CHECKSIGADD` calls to succeed).  This change
+  is made to allow for batch verification of multiple signatures, which
+  [bip-schnorr][] estimates can speed up verification by FIXMEx over
+  checking each signature independently.
+
+    Ideally, many multisig operations can use signature aggregation as
+    described in the multisig subsection above.  However, it may not be
+    obvious that there are many cases where k-of-n multisig can be done
+    using multiple k-of-k scripts each in a separate Taproot branch.
+    For example, consider the following 2-of-3 multisig Tapscript using
+    `OP_CHECKSIGADD` (`OP_CSADD`):
+
+        <A pubkey> OP_CSADD <B pubkey> OP_CSADD <C pubkey> OP_CSADD 2 OP_EQUAL
+
+    This may be more efficient as three different 2-of-2 multisig
+    scripts in a Taproot merkle tree, allowing whichever one is needed
+    to be chosen:
+
+        (1) <A pubkey> OP_CHECKSIGVERIFY <B pubkey> OP_CHECKSIG
+        (2) <A pubkey> OP_CHECKSIGVERIFY <C pubkey> OP_CHECKSIG
+        (3) <B pubkey> OP_CHECKSIGVERIFY <C pubkey> OP_CHECKSIG
+
```

Comment:
Do we want to point out that this will (unfortunately) be heavier weight and more costly than using traditional P2WSH OP_CHECKMULTISIG due to having to broadcast 1 or more MAST hashes?

Reply:
I haven't done the exact math yet, but quick mental math told me the main example in the previous subsection (the HTLC) is going to come out with one of the branches more expensive than P2WSH, and that's why I added a note there, "this means that more
complex scripts than our simple HTLC can gain much larger efficiency
improvements than seen above (see an article focused on [MAST][] for
more information)."

I think being explicit there and showing how small trees can result in higher costs will give us a pass for avoiding details here.  The objective of this example is to show how k-of-n can become multiple k-of-k, and there are many cases for larger values of n where this will be more cost effective than full scripts (plus there's always key-path spending).